File: bindings/java/src/main/java/com/hyperpolymath/proven/SafeMath.java
// SPDX-License-Identifier: PMPL-1.0-or-later
// SPDX-FileCopyrightText: 2025 Hyperpolymath
package com.hyperpolymath.proven;

import java.math.BigInteger;

/**
 * Safe mathematical operations that detect overflow without throwing.
 * All operations return Result types for safe error handling.
 */
public final class SafeMath {
    private SafeMath() {}

    /**
     * Safe addition with overflow detection.
     */
    public static Result<Long, String> addSafe(long a, long b) {
        try {
            return Result.ok(Math.addExact(a, b));
        } catch (ArithmeticException e) {
            return Result.err("Integer overflow in addition");
        }
    }

    /**
     * Safe subtraction with overflow detection.
     */
    public static Result<Long, String> subSafe(long a, long b) {
        try {
            return Result.ok(Math.subtractExact(a, b));
        } catch (ArithmeticException e) {
            return Result.err("Integer overflow in subtraction");
        }
    }

    /**
     * Safe multiplication with overflow detection.
     */
    public static Result<Long, String> mulSafe(long a, long b) {
        try {
            return Result.ok(Math.multiplyExact(a, b));
        } catch (ArithmeticException e) {
            return Result.err("Integer overflow in multiplication");
        }
    }

    /**
     * Safe division with zero check.
     */
    public static Result<Long, String> divSafe(long a, long b) {
        if (b == 0) {
            return Result.err("Division by zero");
        }
        if (a == Long.MIN_VALUE && b == -1) {
            return Result.err("Integer overflow in division");
        }
        return Result.ok(a / b);
    }

    /**
     * Safe modulo with zero check.
     */
    public static Result<Long, String> modSafe(long a, long b) {
        if (b == 0) {
            return Result.err("Modulo by zero");
        }
        return Result.ok(a % b);
    }

    /**
     * Safe negation with overflow detection.
     */
    public static Result<Long, String> negateSafe(long a) {
        try {
            return Result.ok(Math.negateExact(a));
        } catch (ArithmeticException e) {
            return Result.err("Integer overflow in negation");
        }
    }

    /**
     * Safe absolute value with overflow detection.
     */
    public static Result<Long, String> absSafe(long a) {
        if (a == Long.MIN_VALUE) {
            return Result.err("Integer overflow in absolute value");
        }
        return Result.ok(Math.abs(a));
    }

    /**
     * Checked integer addition for int type.
     */
    public static Result<Integer, String> addSafeInt(int a, int b) {
        try {
            return Result.ok(Math.addExact(a, b));
        } catch (ArithmeticException e) {
            return Result.err("Integer overflow in addition");
        }
    }

    /**
     * Checked integer multiplication for int type.
     */
    public static Result<Integer, String> mulSafeInt(int a, int b) {
        try {
            return Result.ok(Math.multiplyExact(a, b));
        } catch (ArithmeticException e) {
            return Result.err("Integer overflow in multiplication");
        }
    }

    /**
     * Safe power operation using BigInteger to detect overflow.
     */
    public static Result<Long, String> powSafe(long base, int exponent) {
        if (exponent < 0) {
            return Result.err("Negative exponent not supported for integer power");
        }
        if (exponent == 0) {
            return Result.ok(1L);
        }
        try {
            BigInteger result = BigInteger.valueOf(base).pow(exponent);
            return Result.ok(result.longValueExact());
        } catch (ArithmeticException e) {
            return Result.err("Integer overflow in power operation");
        }
    }

    /**
     * Clamp a value to a range.
     */
    public static long clamp(long value, long min, long max) {
        return Math.max(min, Math.min(max, value));
    }

    /**
     * Check if addition would overflow.
     */
    public static boolean wouldOverflowAdd(long a, long b) {
        return addSafe(a, b).isErr();
    }

    /**
     * Check if multiplication would overflow.
     */
    public static boolean wouldOverflowMul(long a, long b) {
        return mulSafe(a, b).isErr();
    }
}

File: bindings/kotlin/src/main/kotlin/com/hyperpolymath/proven/SafeMath.kt
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2025 Hyperpolymath

package com.hyperpolymath.proven

/**
 * Safe arithmetic operations that cannot crash or overflow unexpectedly.
 */
object SafeMath {
    /**
     * Safely divide two integers, returning null on division by zero.
     */
    fun div(numerator: Int, denominator: Int): Int? =
        if (denominator == 0) null else numerator / denominator

    /**
     * Safely divide two longs, returning null on division by zero.
     */
    fun div(numerator: Long, denominator: Long): Long? =
        if (denominator == 0L) null else numerator / denominator

    /**
     * Safely compute modulo, returning null on division by zero.
     */
    fun mod(numerator: Int, denominator: Int): Int? =
        if (denominator == 0) null else numerator % denominator

    /**
     * Safely compute modulo for longs, returning null on division by zero.
     */
    fun mod(numerator: Long, denominator: Long): Long? =
        if (denominator == 0L) null else numerator % denominator

    /**
     * Safely add two integers, returning null on overflow.
     */
    fun add(a: Int, b: Int): Int? {
        val result = a.toLong() + b.toLong()
        return if (result > Int.MAX_VALUE || result < Int.MIN_VALUE) null else result.toInt()
    }

    /**
     * Safely add two longs, returning null on overflow.
     */
    fun add(a: Long, b: Long): Long? = try {
        Math.addExact(a, b)
    } catch (e: ArithmeticException) {
        null
    }

    /**
     * Safely subtract two integers, returning null on overflow.
     */
    fun sub(a: Int, b: Int): Int? {
        val result = a.toLong() - b.toLong()
        return if (result > Int.MAX_VALUE || result < Int.MIN_VALUE) null else result.toInt()
    }

    /**
     * Safely subtract two longs, returning null on overflow.
     */
    fun sub(a: Long, b: Long): Long? = try {
        Math.subtractExact(a, b)
    } catch (e: ArithmeticException) {
        null
    }

    /**
     * Safely multiply two integers, returning null on overflow.
     */
    fun mul(a: Int, b: Int): Int? = try {
        Math.multiplyExact(a, b)
    } catch (e: ArithmeticException) {
        null
    }

    /**
     * Safely multiply two longs, returning null on overflow.
     */
    fun mul(a: Long, b: Long): Long? = try {
        Math.multiplyExact(a, b)
    } catch (e: ArithmeticException) {
        null
    }

    /**
     * Result type for checked operations.
     */
    sealed class MathResult<out T> {
        data class Success<T>(val value: T) : MathResult<T>()
        data object Overflow : MathResult<Nothing>()
        data object DivisionByZero : MathResult<Nothing>()
    }

    /**
     * Add with overflow checking, returning a result type.
     */
    fun checkedAdd(a: Long, b: Long): MathResult<Long> = try {
        MathResult.Success(Math.addExact(a, b))
    } catch (e: ArithmeticException) {
        MathResult.Overflow
    }

    /**
     * Subtract with overflow checking, returning a result type.
     */
    fun checkedSub(a: Long, b: Long): MathResult<Long> = try {
        MathResult.Success(Math.subtractExact(a, b))
    } catch (e: ArithmeticException) {
        MathResult.Overflow
    }

    /**
     * Multiply with overflow checking, returning a result type.
     */
    fun checkedMul(a: Long, b: Long): MathResult<Long> = try {
        MathResult.Success(Math.multiplyExact(a, b))
    } catch (e: ArithmeticException) {
        MathResult.Overflow
    }
}
